util/types/json/functions_test.go

```diff
@@ -103,3 +103,75 @@ func (s *testJSONSuite) TestJSONUnquote(c *C) {
 		c.Assert(unquoted, Equals, tt.unquoted)
 	}
 }
+
+func (s *testJSONSuite) TestJSONMerge(c *C) {
+	var tests = []struct {
+		base     string
+		suffixes []string
+		expected string
+	}{
+		{`{"a": 1}`, []string{`{"b": 2}`}, `{"a": 1, "b": 2}`},
+		{`{"a": 1}`, []string{`{"a": 2}`}, `{"a": [1, 2]}`},
+		{`[1]`, []string{`[2]`}, `[1, 2]`},
+		{`{"a": 1}`, []string{`[1]`}, `[{"a": 1}, 1]`},
+		{`[1]`, []string{`{"a": 1}`}, `[1, {"a": 1}]`},
+		{`{"a": 1}`, []string{`4`}, `[{"a": 1}, 4]`},
+		{`[1]`, []string{`4`}, `[1, 4]`},
+		{`4`, []string{`{"a": 1}`}, `[4, {"a": 1}]`},
+		{`4`, []string{`1`}, `[4, 1]`},
+	}
+
+	for _, tt := range tests {
+		base := mustParseFromString(tt.base)
+		suffixes := make([]JSON, 0, len(tt.suffixes))
+		for _, s := range tt.suffixes {
+			suffixes = append(suffixes, mustParseFromString(s))
+		}
+		base = base.Merge(suffixes)
+		cmp, err := CompareJSON(base, mustParseFromString(tt.expected))
+		c.Assert(err, IsNil)
+		c.Assert(cmp, Equals, 0)
+	}
+}
+
+func (s *testJSONSuite) TestJSONModify(c *C) {
+	var tests = []struct {
+		base     string
+		setField string
+		setValue string
+		mt       ModifyType
+		expected string
+	}{
+		{`null`, "$", `{}`, ModifySet, `{}`},
+		{`{}`, "$.a", `3`, ModifySet, `{"a": 3}`},
+		{`{"a": 3}`, "$.a", `[]`, ModifyReplace, `{"a": []}`},
+		{`{"a": []}`, "$.a[0]", `3`, ModifySet, `{"a": [3]}`},
+		{`{"a": [3]}`, "$.a[1]", `4`, ModifyInsert, `{"a": [3, 4]}`},
```

Comment:
This rule applies to `json_extract` too, so some cases in `json_extract` is incorrect.